hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/preloader/NearCachePreloader.java

```diff
@@ -0,0 +1,303 @@
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.util.Set;
+
+import static com.hazelcast.nio.Bits.INT_SIZE_IN_BYTES;
+import static com.hazelcast.nio.Bits.readIntB;
+import static com.hazelcast.nio.Bits.writeIntB;
+import static com.hazelcast.nio.IOUtil.closeResource;
+import static com.hazelcast.nio.IOUtil.deleteQuietly;
+import static com.hazelcast.nio.IOUtil.readFullyOrNothing;
+import static com.hazelcast.nio.IOUtil.rename;
+import static com.hazelcast.nio.IOUtil.toFileName;
+import static com.hazelcast.util.StringUtil.isNullOrEmpty;
+import static java.lang.String.format;
+import static java.nio.ByteBuffer.allocate;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * Loads and stores the keys from a Near Cache into a file.
+ *
+ * @param <KS> type of the {@link NearCacheRecord} keys
+ */
+public class NearCachePreloader<KS> {
+
+    /**
+     * File format for the file header.
+     */
+    private enum FileFormat {
+        INTERLEAVED_LENGTH_FIELD
+    }
+
+    /**
+     * Magic bytes for the file header.
+     */
+    private static final int MAGIC_BYTES = 0xEA3CAC4E;
+
+    /**
+     * Base-2 logarithm of buffer size.
+     */
+    private static final int LOG_OF_BUFFER_SIZE = 16;
+    /**
+     * Buffer size used for file I/O. Invariant: buffer size is a power of two.
+     */
+    private static final int BUFFER_SIZE = 1 << LOG_OF_BUFFER_SIZE;
+
+    /**
+     * Batch size for the pre-loader.
+     */
+    private static final int LOAD_BATCH_SIZE = 100;
+
+    private final ILogger logger = Logger.getLogger(NearCachePreloader.class);
+    private final ByteBuffer buf = allocate(BUFFER_SIZE);
+
+    private final String nearCacheName;
+    private final NearCacheStatsImpl nearCacheStats;
+    private final SerializationService serializationService;
+
+    private final File storeFile;
+    private final File tmpStoreFile;
+
+    private byte[] tmpBytes;
+
+    private int lastWrittenBytes;
+    private int lastKeyCount;
+
+    public NearCachePreloader(String nearCacheName, NearCachePreloaderConfig preloaderConfig,
+                              NearCacheStatsImpl nearCacheStats, SerializationService serializationService) {
+        this.nearCacheName = nearCacheName;
+        this.nearCacheStats = nearCacheStats;
+        this.serializationService = serializationService;
+
+        String fileName = getFileName(preloaderConfig.getFileName(), nearCacheName);
+        this.storeFile = new File(fileName);
+        this.tmpStoreFile = new File(fileName + "~");
+    }
+
+    /**
+     * Loads the values via a stored key file into the supplied {@link DataStructureAdapter}.
+     *
+     * @param adapter the {@link DataStructureAdapter} to load the values from
+     */
+    public void loadKeys(DataStructureAdapter<Data, ?> adapter) {
+        long started = System.nanoTime();
```

Comment:
I would name this startedNanos. This is because in a single method you make use of 2 time units; nanos; millis.